benchmarks-mobile/protolite_app/app/src/main/java/io/grpc/grpcbenchmarks/RpcBenchmarksActivity.java

```diff
@@ -1 +1,205 @@
+            });
+            cardViews.add(cv);
+            l.addView(cv);
+        }
+    }
+
+    public void startBenchmark(CardView cv, RpcBenchmark b) {
+        String host = mHostEditText.getText().toString();
+        String payloadSize = mPayloadEditText.getText().toString();
+        String useGzip = Boolean.toString(mGzip.isChecked());
+        String useOkHttp = Boolean.toString(mOkHttp.isChecked());
+
+        // set default payload size
+        if (payloadSize.length() == 0) {
+            payloadSize = "100";
+        }
+
+        BenchmarkAsyncTask task = new BenchmarkAsyncTask(cv, b);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
+            task.executeOnExecutor(AsyncTask.SERIAL_EXECUTOR, host, payloadSize,
+                    useGzip, useOkHttp);
+        } else {
+            task.execute(host, payloadSize, useGzip, useOkHttp);
+        }
+    }
+
+    public void pingAddress(View v) {
+        String host = mHostEditText.getText().toString();
+        new PingAsyncTask().execute(host);
+    }
+
+    public void beginAllBenchmarks(View v) {
+        mBenchmarkButton.setEnabled(false);
+        mBenchmarkButton.setText(R.string.allBenchmarksButtonDisabled);
+        for (CardView cv : cardViews) {
+            cv.findViewById(R.id.protobuf_benchmark_start).performClick();
+        }
+    }
+
+    private class PingAsyncTask extends AsyncTask<String, Void, String> {
+        @Override
+        protected void onPreExecute() {
+            mPingTextView.setText(R.string.pingRunning);
+            mPingButton.setEnabled(false);
+        }
+
+        @Override
+        protected String doInBackground(String... args) {
+            try {
+                Process p = Runtime.getRuntime().exec(new String[]{"ping", "-c", "4", args[0]});
+                BufferedReader stdInput = new BufferedReader(new InputStreamReader(p.getInputStream()));
+
+                String s;
+                int count = 0;
+                while ((s = stdInput.readLine()) != null) {
+                    count++;
+                    // 4 + 5 to get the last line
+                    if (count == 9) {
+                        return s.trim();
+                    }
+                }
+            } catch (Exception e) {
+                logger.log(Level.WARNING, "Failed to ping " + args[0]);
+            }
+            return "Failed to ping host, is server running/reachable?";
+        }
+
+        @Override
+        protected void onPostExecute(String result) {
+            mPingButton.setEnabled(true);
+            mPingTextView.setText(result);
+        }
+    }
+
+    private class BenchmarkAsyncTask extends AsyncTask<String, Void, RpcBenchmarkResult> {
+        CardView cv;
+        RpcBenchmark b;
+
+        BenchmarkAsyncTask(CardView cv, RpcBenchmark b) {
+            this.cv = cv;
+            this.b = b;
+        }
+
+        @Override
+        protected void onPreExecute() {
+            tasksRunning++;
```

Comment:
Do we really need to allow concurrent benchmarks? It would affect the accuracy of the results.